File: app/src/main/java/vinhlong/ditagis/com/qlcln/async/EditAsync.java
package vinhlong.ditagis.com.qlcln.async;

import android.app.ProgressDialog;
import android.content.Context;
import android.os.AsyncTask;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.ArcGISFeature;
import com.esri.arcgisruntime.data.CodedValue;
import com.esri.arcgisruntime.data.CodedValueDomain;
import com.esri.arcgisruntime.data.Domain;
import com.esri.arcgisruntime.data.FeatureEditResult;
import com.esri.arcgisruntime.data.FeatureType;
import com.esri.arcgisruntime.data.ServiceFeatureTable;

import java.text.ParseException;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.concurrent.ExecutionException;

import vinhlong.ditagis.com.qlcln.R;
import vinhlong.ditagis.com.qlcln.adapter.FeatureViewMoreInfoAdapter;
import vinhlong.ditagis.com.qlcln.utities.Constant;

/**
 * Created by ThanLe on 4/16/2018.
 */

public class EditAsync extends AsyncTask<FeatureViewMoreInfoAdapter, Object, Void> {
    private ProgressDialog dialog;
    private Context mContext;
    private ServiceFeatureTable mServiceFeatureTable;
    private ArcGISFeature mSelectedArcGISFeature = null;
    private AsyncResponse mDelegate;

   public interface AsyncResponse {
        void processFinish(Object o);
    }

    public EditAsync(Context context, ServiceFeatureTable serviceFeatureTable, ArcGISFeature selectedArcGISFeature,
                     AsyncResponse response) {
        mContext = context;
        mServiceFeatureTable = serviceFeatureTable;
        mSelectedArcGISFeature = selectedArcGISFeature;
        dialog = new ProgressDialog(context, android.R.style.Theme_Material_Dialog_Alert);
        mDelegate = response;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        dialog.setMessage("Đang cập nhật...");
        dialog.setCancelable(false);

        dialog.show();

    }

    @Override
    protected Void doInBackground(FeatureViewMoreInfoAdapter... params) {
        FeatureViewMoreInfoAdapter adapter = params[0];
        for (FeatureViewMoreInfoAdapter.Item item : adapter.getItems()) {
            if (item.getValue() == null) continue;
            Domain domain = mSelectedArcGISFeature.getFeatureTable().getField(item.getFieldName()).getDomain();
            Object codeDomain = null;
            if (domain != null) {
                List<CodedValue> codedValues = ((CodedValueDomain) this.mSelectedArcGISFeature.getFeatureTable().getField(item.getFieldName()).getDomain()).getCodedValues();
                codeDomain = getCodeDomain(codedValues, item.getValue());
            }
            if (item.getFieldName().equals(mSelectedArcGISFeature.getFeatureTable().getTypeIdField())) {
                List<FeatureType> featureTypes = mSelectedArcGISFeature.getFeatureTable().getFeatureTypes();
                Object idFeatureTypes = getIdFeatureTypes(featureTypes, item.getValue());
                mSelectedArcGISFeature.getAttributes().put(item.getFieldName(), Short.parseShort(idFeatureTypes.toString()));

            } else switch (item.getFieldType()) {
                case DATE:
                    Date date = null;
                    try {
                        date = Constant.DATE_FORMAT.parse(item.getValue());
                        Calendar c = Calendar.getInstance();
                        c.setTime(date);
                        mSelectedArcGISFeature.getAttributes().put(item.getFieldName(), c);
                    } catch (ParseException e) {
                    }
                    break;

                case TEXT:
                    if (codeDomain != null) {
                        mSelectedArcGISFeature.getAttributes().put(item.getFieldName(), codeDomain.toString());
                    } else
                        mSelectedArcGISFeature.getAttributes().put(item.getFieldName(), item.getValue());
                    break;
                case SHORT:
                    if (codeDomain != null) {
                        mSelectedArcGISFeature.getAttributes().put(item.getFieldName(), Short.parseShort(codeDomain.toString()));
                    } else
                        mSelectedArcGISFeature.getAttributes().put(item.getFieldName(), Short.parseShort(item.getValue()));
                    break;
            }
        }
        Calendar currentTime = Calendar.getInstance();
        mSelectedArcGISFeature.getAttributes().put("NgayCapNhat", currentTime);
        mServiceFeatureTable.loadAsync();
        mServiceFeatureTable.addDoneLoadingListener(new Runnable() {
            @Override
            public void run() {
                try {
                    // update feature in the feature table
                    mServiceFeatureTable.updateFeatureAsync(mSelectedArcGISFeature).addDoneListener(new Runnable() {
                        @Override
                        public void run() {
                            ListenableFuture<List<FeatureEditResult>> applyEditsAsync = mServiceFeatureTable.applyEditsAsync();
                            applyEditsAsync.addDoneListener(new Runnable() {
                                @Override
                                public void run() {
                                    try {
                                        List<FeatureEditResult> featureEditResults = applyEditsAsync.get();
                                        if (featureEditResults.size() > 0 && !featureEditResults.get(0).hasCompletedWithErrors()){
                                            publishProgress(featureEditResults.get(0).getObjectId());
                                        }
                                        else publishProgress();
                                    } catch (Exception e) {
                                        publishProgress();
                                    }

                                }
                            });
                        }
                    });

                } catch (Exception e) {
                }
            }
        });
        return null;
    }

    private Object getIdFeatureTypes(List<FeatureType> featureTypes, String value) {
        Object code = null;
        for (FeatureType featureType : featureTypes) {
            if (featureType.getName().equals(value)) {
                code = featureType.getId();
                break;
            }
        }
        return code;
    }

    private Object getCodeDomain(List<CodedValue> codedValues, String value) {
        Object code = null;
        for (CodedValue codedValue : codedValues) {
            if (codedValue.getName().equals(value)) {
                code = codedValue.getCode();
                break;
            }

        }
        return code;
    }

    @Override
    protected void onProgressUpdate(Object... values) {
        super.onProgressUpdate(values);
        if (values != null && values.length > 0) {
            mDelegate.processFinish(values[0]);
        } else {
            mDelegate.processFinish(false);
        }
        if (dialog != null && dialog.isShowing()) {
            dialog.dismiss();
        }
    }


}

File: app/src/main/java/vinhlong/ditagis/com/qlcln/async/EditAsync.kt
package vinhlong.ditagis.com.qlcln.async

import android.app.ProgressDialog
import android.content.Context
import android.os.AsyncTask
import com.esri.arcgisruntime.data.*
import vinhlong.ditagis.com.qlcln.adapter.FeatureViewMoreInfoAdapter
import vinhlong.ditagis.com.qlcln.utities.Constant
import java.text.ParseException
import java.util.*

/**
 * Created by ThanLe on 4/16/2018.
 */

class EditAsync(private val mContext: Context, private val mServiceFeatureTable: ServiceFeatureTable, selectedArcGISFeature: ArcGISFeature,
                private val mDelegate: AsyncResponse) : AsyncTask<FeatureViewMoreInfoAdapter, Any, Void?>() {
    private val dialog: ProgressDialog?
    private var mSelectedArcGISFeature: ArcGISFeature? = null

    interface AsyncResponse {
        fun processFinish(o: Any)
    }

    init {
        mSelectedArcGISFeature = selectedArcGISFeature
        dialog = ProgressDialog(mContext, android.R.style.Theme_Material_Dialog_Alert)
    }

    override fun onPreExecute() {
        super.onPreExecute()
        dialog!!.setMessage("Đang cập nhật...")
        dialog.setCancelable(false)

        dialog.show()

    }

    override fun doInBackground(vararg params: FeatureViewMoreInfoAdapter): Void? {
        val adapter = params[0]
        for (item in adapter.getItems()) {
            if (item.value == null) continue
            val domain = mSelectedArcGISFeature!!.featureTable.getField(item.fieldName!!).domain
            var codeDomain: Any? = null
            if (domain != null) {
                val codedValues = (this.mSelectedArcGISFeature!!.featureTable.getField(item.fieldName!!).domain as CodedValueDomain).codedValues
                codeDomain = getCodeDomain(codedValues, item.value)
            }
            if (item.fieldName == mSelectedArcGISFeature!!.featureTable.typeIdField) {
                val featureTypes = mSelectedArcGISFeature!!.featureTable.featureTypes
                val idFeatureTypes = getIdFeatureTypes(featureTypes, item.value)
                mSelectedArcGISFeature!!.attributes[item.fieldName] = java.lang.Short.parseShort(idFeatureTypes!!.toString())

            } else
                when (item.fieldType) {
                    Field.Type.DATE -> {
                        var date: Date? = null
                        try {
                            date = Constant.DATE_FORMAT.parse(item.value)
                            val c = Calendar.getInstance()
                            c.time = date
                            mSelectedArcGISFeature!!.attributes[item.fieldName] = c
                        } catch (e: ParseException) {
                        }

                    }

                    Field.Type.TEXT -> if (codeDomain != null) {
                        mSelectedArcGISFeature!!.attributes[item.fieldName] = codeDomain.toString()
                    } else
                        mSelectedArcGISFeature!!.attributes[item.fieldName] = item.value
                    Field.Type.SHORT -> if (codeDomain != null) {
                        mSelectedArcGISFeature!!.attributes[item.fieldName] = java.lang.Short.parseShort(codeDomain.toString())
                    } else
                        mSelectedArcGISFeature!!.attributes[item.fieldName] = java.lang.Short.parseShort(item.value)
                }
        }
        val currentTime = Calendar.getInstance()
        mSelectedArcGISFeature!!.attributes["NgayCapNhat"] = currentTime
        mServiceFeatureTable.loadAsync()
        mServiceFeatureTable.addDoneLoadingListener {
            try {
                // update feature in the feature table
                mServiceFeatureTable.updateFeatureAsync(mSelectedArcGISFeature!!).addDoneListener {
                    val applyEditsAsync = mServiceFeatureTable.applyEditsAsync()
                    applyEditsAsync.addDoneListener {
                        try {
                            val featureEditResults = applyEditsAsync.get()
                            if (featureEditResults.size > 0 && !featureEditResults[0].hasCompletedWithErrors()) {
                                publishProgress(featureEditResults[0].objectId)
                            } else
                                publishProgress()
                        } catch (e: Exception) {
                            publishProgress()
                        }
                    }
                }

            } catch (e: Exception) {
            }
        }
        return null
    }

    private fun getIdFeatureTypes(featureTypes: List<FeatureType>, value: String?): Any? {
        var code: Any? = null
        for (featureType in featureTypes) {
            if (featureType.name == value) {
                code = featureType.id
                break
            }
        }
        return code
    }

    private fun getCodeDomain(codedValues: List<CodedValue>, value: String?): Any? {
        var code: Any? = null
        for (codedValue in codedValues) {
            if (codedValue.name == value) {
                code = codedValue.code
                break
            }

        }
        return code
    }

    override fun onProgressUpdate(vararg values: Any) {
        super.onProgressUpdate(*values)
        if (values.isNotEmpty()) {
            mDelegate.processFinish(values[0])
        } else {
            mDelegate.processFinish(false)
        }
        if (dialog != null && dialog.isShowing) {
            dialog.dismiss()
        }
    }


}

